File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/PhysicalLogRecord.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Fabric;
    using System.IO;

    internal class PhysicalLogRecord : LogRecord
    {
        internal static readonly PhysicalLogRecord InvalidPhysicalLogRecord = new PhysicalLogRecord();

        private static readonly uint ApproximateDiskSpaceUsed = sizeof(int) + sizeof(ulong) + sizeof(long);

        // The following fields are not persisted
        private PhysicalLogRecord linkedPhysicalRecord;

        private ulong linkedPhysicalRecordOffset;

        private WeakReference<PhysicalLogRecord> nextPhysicalRecord;

        protected PhysicalLogRecord(LogRecordType recordType, ulong recordPosition, long lsn)
            : base(recordType, recordPosition, lsn)
        {
            this.linkedPhysicalRecordOffset = InvalidPhysicalRecordOffset;
            this.linkedPhysicalRecord = InvalidPhysicalLogRecord;
            this.nextPhysicalRecord = new WeakReference<PhysicalLogRecord>(InvalidPhysicalLogRecord);
        }

        protected PhysicalLogRecord(
            LogRecordType recordType,
            LogicalSequenceNumber lsn,
            PhysicalLogRecord linkedPhysicalRecord)
            : base(recordType)
        {
            this.Lsn = lsn;
            if (linkedPhysicalRecord != null)
            {
                this.linkedPhysicalRecordOffset = InvalidPhysicalRecordOffset;
                this.linkedPhysicalRecord = linkedPhysicalRecord;
            }
            else
            {
                this.linkedPhysicalRecordOffset = 0;
                this.linkedPhysicalRecord = null;
            }

            this.nextPhysicalRecord = new WeakReference<PhysicalLogRecord>(InvalidPhysicalLogRecord);

            this.UpdateApproximateDiskSize();
        }

        protected PhysicalLogRecord()
            : base()
        {
            this.linkedPhysicalRecordOffset = InvalidPhysicalRecordOffset;
            this.linkedPhysicalRecord = null;
            this.nextPhysicalRecord = null;
        }

        internal PhysicalLogRecord LinkedPhysicalRecord
        {
            get { return this.linkedPhysicalRecord; }

            set { this.linkedPhysicalRecord = value; }
        }

        internal ulong LinkedPhysicalRecordOffset
        {
            get { return this.linkedPhysicalRecordOffset; }

            // Test only set.
            set { this.linkedPhysicalRecordOffset = value; }
        }

        internal PhysicalLogRecord NextPhysicalRecord
        {
            get
            {
                if (this.nextPhysicalRecord == null)
                {
                    return null;
                }

                PhysicalLogRecord result = null;
                bool exists = this.nextPhysicalRecord.TryGetTarget(out result);

#if (!DEBUG)
                Utility.Assert(
                    exists,
                    "NextPhysicalRecord for Lsn: {0} is not present",
                    this.Lsn);
#endif

                return result;
            }

            set { this.nextPhysicalRecord = new WeakReference<PhysicalLogRecord>(value); }
        }
        
        protected override void Read(BinaryReader br, bool isPhysicalRead)
        {
            base.Read(br, isPhysicalRead);

            var startingPosition = br.BaseStream.Position;
            var sizeOfSection = br.ReadInt32();
            var endPosition = startingPosition + sizeOfSection;

            this.linkedPhysicalRecordOffset = br.ReadUInt64();

            // Jump to the end of the section ignoring fields that are not understood.
            Utility.Assert(endPosition >= br.BaseStream.Position, "Could not have read more than section size.");
            br.BaseStream.Position = endPosition;

            this.UpdateApproximateDiskSize();
            return;
        }

        protected override void Write(BinaryWriter bw, OperationData operationData, bool isPhysicalWrite, bool forceRecomputeOffsets)
        {
            base.Write(bw, operationData, isPhysicalWrite, forceRecomputeOffsets);
            var startingPosition = bw.BaseStream.Position;
            bw.BaseStream.Position += sizeof(int);

            if (this.linkedPhysicalRecordOffset == InvalidPhysicalRecordOffset || forceRecomputeOffsets == true)
            {
                Utility.Assert(
                    this.linkedPhysicalRecord != null || forceRecomputeOffsets == true,
                    "(this.linkedPhysicalRecord != null)={0}, forceRecomputeOffsets={1}",
                    this.linkedPhysicalRecord != null,
                    forceRecomputeOffsets);

                if (this.linkedPhysicalRecord == null)
                {
                    this.linkedPhysicalRecordOffset = 0;
                }
                else
                {
                    Utility.Assert(
                        this.linkedPhysicalRecord.RecordPosition != InvalidRecordPosition,
                        "this.linkedPhysicalRecord.RecordPosition != LogRecord.INVALID_RECORD_POSITION");
                    Utility.Assert(
                        this.RecordPosition != InvalidRecordPosition,
                        "this.RecordPosition != LogRecord.INVALID_RECORD_POSITION");

                    this.linkedPhysicalRecordOffset = this.RecordPosition - this.linkedPhysicalRecord.RecordPosition;
                }
            }

            bw.Write(this.linkedPhysicalRecordOffset);

            var endPosition = bw.BaseStream.Position;
            var sizeOfSection = checked((int) (endPosition - startingPosition));
            bw.BaseStream.Position = startingPosition;
            bw.Write(sizeOfSection);
            bw.BaseStream.Position = endPosition;

            operationData.Add(CreateArraySegment(startingPosition, bw));
            return;
        }

        internal override bool FreePreviousLinksLowerThanPsn(PhysicalSequenceNumber newHeadPsn)
        {
            bool ret = base.FreePreviousLinksLowerThanPsn(newHeadPsn);

            if (this.linkedPhysicalRecord != null &&
                this.linkedPhysicalRecord.Psn < newHeadPsn)
            {
                Utility.Assert(
                    this.linkedPhysicalRecordOffset != InvalidPhysicalRecordOffset,
                    "FreePreviousLinksLowerThanPsn: PreviousPhysicalRecordOffset is invalid. Record LSN is {0}");

                this.linkedPhysicalRecord = (this.linkedPhysicalRecordOffset == 0)
                    ? null
                    : PhysicalLogRecord.InvalidPhysicalLogRecord;

                return true;
            }

            return ret;
        }

        private void UpdateApproximateDiskSize()
        {
            this.ApproximateSizeOnDisk += ApproximateDiskSpaceUsed;
        }
    }
}

File: src/prod/src/data/txnreplicator/logrecordlib/PhysicalLogRecord.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace Data::LogRecordLib;
using namespace TxnReplicator;
using namespace Data::Utilities;

const ULONG PhysicalLogRecord::DiskSpaceUsed = sizeof(LONG32) + sizeof(ULONG64);

PhysicalLogRecord::PhysicalLogRecord()
    : LogRecord()
    , KWeakRefType()
    , linkedPhysicalRecordOffset_(Constants::InvalidPhysicalRecordOffset)
    , linkedPhysicalRecord_(nullptr)
    , nextPhysicalRecord_(nullptr)
{
}

PhysicalLogRecord::PhysicalLogRecord(
    __in LogRecordType::Enum recordType,
    __in ULONG64 recordPosition,
    __in LONG64 lsn,
    __in PhysicalLogRecord & invalidPhysicalLogRecord)
    : LogRecord(recordType, recordPosition, lsn, invalidPhysicalLogRecord)
    , KWeakRefType()
    , linkedPhysicalRecordOffset_(Constants::InvalidPhysicalRecordOffset)
    , linkedPhysicalRecord_(&invalidPhysicalLogRecord)
    , nextPhysicalRecord_(invalidPhysicalLogRecord.GetWeakRef())
{
    ASSERT_IFNOT(
        invalidPhysicalLogRecord.linkedPhysicalRecordOffset_ == Constants::InvalidPhysicalRecordOffset,
        "Invalid physical record cannot have a valid linked physical record offset");
}

PhysicalLogRecord::PhysicalLogRecord(
    __in LogRecordType::Enum recordType, 
    __in LONG64 lsn, 
    __in_opt PhysicalLogRecord * const linkedPhysicalRecord,
    __in PhysicalLogRecord & invalidPhysicalLogRecord)
    : LogRecord(recordType, invalidPhysicalLogRecord)
    , KWeakRefType()
    , linkedPhysicalRecordOffset_(0)
    , linkedPhysicalRecord_(linkedPhysicalRecord)
    , nextPhysicalRecord_(invalidPhysicalLogRecord.GetWeakRef())
{
    ASSERT_IFNOT(
        linkedPhysicalRecord == nullptr ||
        linkedPhysicalRecord->RecordType == LogRecordType::Enum::EndCheckpoint ||
        linkedPhysicalRecord->RecordType == LogRecordType::Enum::CompleteCheckpoint ||
        linkedPhysicalRecord->RecordType == LogRecordType::Enum::TruncateHead,
        "Invalid linked physical record. Acceptable values are EndCheckpointLogRecord, CompleteCheckpointLogRecord, TruncateHeadLogRecord or nullptr");

    Lsn = lsn;

    if (linkedPhysicalRecord_ != nullptr)
    {
        linkedPhysicalRecordOffset_ = Constants::InvalidPhysicalRecordOffset;
    }

    ASSERT_IFNOT(
        invalidPhysicalLogRecord.linkedPhysicalRecordOffset_ == Constants::InvalidPhysicalRecordOffset,
        "Invalid physical record cannot have a valid linked physical record offset");

    UpdateApproximateDiskSize();
}

PhysicalLogRecord::~PhysicalLogRecord()
{
    FreePhysicalLinkChainProactively();
}

void PhysicalLogRecord::UpdateApproximateDiskSize()
{
    ApproximateSizeOnDisk = ApproximateSizeOnDisk + DiskSpaceUsed;
}

void PhysicalLogRecord::Read(
    __in BinaryReader & binaryReader, 
    __in bool isPhysicalRead)
{
    __super::Read(binaryReader, isPhysicalRead);
    ULONG32 startingPosition = binaryReader.Position;
    ULONG32 sizeOfSection;
    binaryReader.Read(sizeOfSection);

    ULONG32 endPosition = startingPosition + sizeOfSection;
    binaryReader.Read(linkedPhysicalRecordOffset_);

    // Jump to the end of the section ignoring fields that are not understood.
    ASSERT_IFNOT(endPosition >= binaryReader.Position,
        "Endposition: {0} should be greater than binary reader position: {1} reading physical log record",
        endPosition, binaryReader.Position);

    binaryReader.Position = endPosition;

    UpdateApproximateDiskSize();
}

void PhysicalLogRecord::Write(
    __in BinaryWriter & binaryWriter, 
    __in OperationData & operationData,
    __in bool isPhysicalWrite,
    __in bool forceRecomputeOffsets)
{
    __super::Write(binaryWriter, operationData, isPhysicalWrite, forceRecomputeOffsets);
    ULONG32 startingPosition = binaryWriter.Position;
    binaryWriter.Position += sizeof(ULONG32);

    if(linkedPhysicalRecordOffset_ == Constants::InvalidPhysicalRecordOffset || forceRecomputeOffsets == true)
    {
        ASSERT_IFNOT(
            linkedPhysicalRecord_ != nullptr || forceRecomputeOffsets == true,
            "Linked physical record must not be null. (linkedPhysicalRecord_ != nullptr)={0}. forceRecomputeOffsets={1}",
            linkedPhysicalRecord_ != nullptr,
            forceRecomputeOffsets);

        if (linkedPhysicalRecord_ == nullptr)
        {
            linkedPhysicalRecordOffset_ = 0;
        }
        else
        {
            ASSERT_IFNOT(
                linkedPhysicalRecord_->RecordPosition != Constants::InvalidRecordPosition,
                "Linked physical record position must be valid");

            ASSERT_IFNOT(
                RecordPosition != Constants::InvalidRecordPosition,
                "Record position must be valid");

            linkedPhysicalRecordOffset_ = RecordPosition - linkedPhysicalRecord_->RecordPosition;
        }
    }

    binaryWriter.Write(linkedPhysicalRecordOffset_);
    ULONG32 endPosition = binaryWriter.Position;
    ULONG32 sizeOfSection = endPosition - startingPosition;

    binaryWriter.Position = startingPosition;
    binaryWriter.Write(sizeOfSection);
    binaryWriter.Position = endPosition;

    KBuffer::SPtr kBuffer = binaryWriter.GetBuffer(startingPosition);
    operationData.Append(*kBuffer);
}

std::wstring PhysicalLogRecord::ToString() const
{
    return __super::ToString();
}

bool PhysicalLogRecord::FreePreviousLinksLowerThanPsn(
    __in LONG64 logHeadPsn,
    __in InvalidLogRecords & invalidLogRecords)
{
    bool ret = __super::FreePreviousLinksLowerThanPsn(
        logHeadPsn, 
        invalidLogRecords);

    if (linkedPhysicalRecord_ != nullptr &&
        linkedPhysicalRecord_->Psn < logHeadPsn)
    {
        ASSERT_IF(
            linkedPhysicalRecordOffset_ == Constants::InvalidPhysicalRecordOffset,
            "FreePreviousLinksLowerThanPsn: Invalid linked physical record offset: {0}",
            linkedPhysicalRecordOffset_);

        linkedPhysicalRecord_ = linkedPhysicalRecordOffset_ == 0 ?
            nullptr :
            invalidLogRecords.Inv_PhysicalLogRecord;

        return true;
    }

    return ret;
}

void PhysicalLogRecord::FreePhysicalLinkChainProactively()
{
    // To prevent a stack overflow, release the back references in a loop proactively
    // Using raw pointers to not increment refcount as it is needed for the checks below
    PhysicalLogRecord * parent = previousPhysicalRecord_.RawPtr();
    PhysicalLogRecord * child = previousPhysicalRecord_.RawPtr();

    // Step 1- go back all the way to the first physical record 
    // Stop if parent is INVALID or null
    while (
        parent != nullptr && 
        parent->RecordPosition != Constants::InvalidRecordPosition)
    {
        child = parent;
        parent = parent->PreviousPhysicalRecord.RawPtr();
    }

    // Step 2 - Free the parent one by one 
    while (parent != previousPhysicalRecord_.RawPtr())
    {
        parent = child;
        child = child->NextPhysicalRecord.RawPtr();
        parent->previousPhysicalRecord_ = nullptr;
    }
}

bool PhysicalLogRecord::Test_Equals(__in LogRecord const & other) const
{
    PhysicalLogRecord const & otherPhysicalRecord = dynamic_cast<PhysicalLogRecord const &>(other);

    if (__super::Test_Equals(other))
    {
        return linkedPhysicalRecordOffset_ == otherPhysicalRecord.linkedPhysicalRecordOffset_;
    }

    return false;
}
